[mlir][gpu] Add 'cluster_stride' attribute to gpu.subgroup_reduce

andfau-amd opened the pull request:
Follow-up to 7aa22f013e24d20291aad745368ff907baa9dfa4, adding an additional attribute needed in some applications.

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -190,12 +218,9 @@ struct ScalarSubgroupReduceToShuffles final
-             << " is greater than subgroup size " << subgroupSize;
-    unsigned effectiveClusterSize = clusterSize.value_or(subgroupSize);
+    ClusterInfo clusterInfo;
+    if (clusterInfo.getAndValidate(op, subgroupSize).failed())
```
prefer free functions for succeeded / failed

andfau-amd commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
```diff
@@ -1230,27 +1233,38 @@ def GPU_SubgroupReduceOp : GPU_Op<"subgroup_reduce", [SameOperandsAndResultType]
+               "std::optional<uint32_t>":$cluster_size,
+               "uint32_t":$cluster_stride), [{
+      build($_builder, $_state, value, op, uniform,
+            cluster_size ? $_builder.getI32IntegerAttr(*cluster_size) : nullptr,
```
I'm using it in the lowerings that break down a single subgroup reduction into several (the bits higher up in `SubgroupReduceLowering.cpp`), where I want to preserve the original cluster attributes or lack thereof. So it is actually used.

andfau-amd commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,44 +137,75 @@ struct ScalarizeSingleElementReduce final
+  unsigned clusterStride;
+  unsigned clusterSize;
+  unsigned subgroupSize;
+  LogicalResult getAndValidate(gpu::SubgroupReduceOp op,
+                               unsigned subgroupSize) {
```
I refactored the code around this struct because I felt it was getting a bit unwieldy passing three unsigned integer parameters around, and I wanted to share the logic that fetches them from the op and validates them between the two users. Though there are only two users of that part, so I'm not sure if I really made the right choice there.

I originally tried to make it a free function, but keeping `clusterStride`, `clusterSize` and `subgroupSize` as variables in scope was more convenient. Maybe it's a bit too weird?

andfau-amd replied:
I've reworked this now, hopefully it's more to your liking. I do think the new way is probably clearer.

andfau-amd commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,44 +137,75 @@ struct ScalarizeSingleElementReduce final
+                                 gpu::AllReduceOperation mode,
+                                 function_ref<Value(Value)> packFn,
+                                 function_ref<Value(Value)> unpackFn) const {
+    assert(llvm::isPowerOf2_32(clusterStride));
+    assert(llvm::isPowerOf2_32(clusterSize));
+    assert(llvm::isPowerOf2_32(subgroupSize));
+    assert(clusterStride < subgroupSize);
+    assert(clusterSize <= subgroupSize);
```
Ah, now that this is a method, you might be right. I'll rework it.

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,34 +137,67 @@ struct ScalarizeSingleElementReduce final
+  unsigned subgroupSize;
+};
+
+static LogicalResult getAndValidateClusterInfo(ClusterInfo &out,
```
Could we return `FailureOr<ClusterInfo>` instead of having it as an out parameter?

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,34 +137,67 @@ struct ScalarizeSingleElementReduce final
+
+  std::optional<uint32_t> clusterSize = op.getClusterSize();
+  assert(!clusterSize ||
+         llvm::isPowerOf2_32(*clusterSize)); // verifier should've caught this
```
```suggestion
         llvm::isPowerOf2_32(*clusterSize)); // Verifier should've caught this.
```

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,34 +137,67 @@ struct ScalarizeSingleElementReduce final
+    return op.emitOpError()
+           << "cluster size " << *clusterSize
+           << " is greater than subgroup size " << subgroupSize;
+  out.clusterSize = clusterSize.value_or(subgroupSize); // effective size
```
```suggestion
  out.clusterSize = clusterSize.value_or(subgroupSize); // Effective size.
```

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,34 +137,67 @@ struct ScalarizeSingleElementReduce final
+  out.clusterSize = clusterSize.value_or(subgroupSize); // effective size
+
+  auto clusterStride = op.getClusterStride();
+  assert(llvm::isPowerOf2_32(clusterStride)); // verifier should've caught this
```
```suggestion
  assert(llvm::isPowerOf2_32(clusterStride)); // Verifier should've caught this.
```

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,34 +137,68 @@ struct ScalarizeSingleElementReduce final
+
+  std::optional<uint32_t> clusterSize = op.getClusterSize();
+  assert(!clusterSize ||
+         llvm::isPowerOf2_32(*clusterSize)); // Verifier should've caught this
```
```suggestion
         llvm::isPowerOf2_32(*clusterSize)); // Verifier should've caught this.
```

See https://llvm.org/docs/CodingStandards.html#commenting

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,34 +137,68 @@ struct ScalarizeSingleElementReduce final
+    return op.emitOpError()
+           << "cluster size " << *clusterSize
+           << " is greater than subgroup size " << subgroupSize;
+  res.clusterSize = clusterSize.value_or(subgroupSize); // Effective size
```
```suggestion
  res.clusterSize = clusterSize.value_or(subgroupSize); // Effective size.
```

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,34 +137,68 @@ struct ScalarizeSingleElementReduce final
+  res.clusterSize = clusterSize.value_or(subgroupSize); // Effective size
+
+  auto clusterStride = op.getClusterStride();
+  assert(llvm::isPowerOf2_32(clusterStride)); // Verifier should've caught this
```
```suggestion
  assert(llvm::isPowerOf2_32(clusterStride)); // Verifier should've caught this.
```

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,34 +137,68 @@ struct ScalarizeSingleElementReduce final
+
+static FailureOr<ClusterInfo>
+getAndValidateClusterInfo(gpu::SubgroupReduceOp op, unsigned subgroupSize) {
+  ClusterInfo res;
```
Let's make sure there's no uninitialized data here

```suggestion
  ClusterInfo res = {};
```

kuhar reviewed (approved):
LGTM % nits

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -140,34 +137,65 @@ struct ScalarizeSingleElementReduce final
+           << "cluster stride " << clusterStride
+           << " is not less than subgroup size " << subgroupSize;
+
+  return success(
+      ClusterInfo{clusterStride, effectiveClusterSize, subgroupSize});
```
```suggestion
  return ClusterInfo{clusterStride, effectiveClusterSize, subgroupSize};
```

kuhar reviewed (commented):
One last nit

andfau-amd committed:
[mlir][gpu] Add 'cluster_stride' attribute to gpu.subgroup_reduce

Follow-up to 7aa22f013e24d20291aad745368ff907baa9dfa4, adding an
additional attribute needed in some applications.

kuhar reviewed (approved):
LGTM
